Fix apt-get error on conf file conflict

jasisk opened the pull request:
During provisioning, if there exists a `docker.conf`, you're prompted to perform an action regardless of the noninteractive and `-y` flags.

I've added a flag which forces the prompt to use the old config which, in the case of provisioning a new machine, should prove completely sufficient as it's likely another package being installed that dropped the config in the first place.

Here's a bunch of noise:

```
Selecting previously unselected package liberror-perl.
(Reading database ... 158291 files and directories currently installed.)
Preparing to unpack .../liberror-perl_0.17-1.1_all.deb ...
Unpacking liberror-perl (0.17-1.1) ...
Selecting previously unselected package git-man.
Preparing to unpack .../git-man_1%3a1.9.1-1ubuntu0.1_all.deb ...
Unpacking git-man (1:1.9.1-1ubuntu0.1) ...
Selecting previously unselected package git.
Preparing to unpack .../git_1%3a1.9.1-1ubuntu0.1_amd64.deb ...
Unpacking git (1:1.9.1-1ubuntu0.1) ...
Selecting previously unselected package cgroup-lite.
Preparing to unpack .../cgroup-lite_1.9_all.deb ...
Unpacking cgroup-lite (1.9) ...
Selecting previously unselected package lxc-docker-1.7.1.
Preparing to unpack .../lxc-docker-1.7.1_1.7.1_amd64.deb ...
Unpacking lxc-docker-1.7.1 (1.7.1) ...
Selecting previously unselected package lxc-docker.
Preparing to unpack .../lxc-docker_1.7.1_amd64.deb ...
Unpacking lxc-docker (1.7.1) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Processing triggers for ureadahead (0.100.0-16) ...
Setting up liberror-perl (0.17-1.1) ...
Setting up git-man (1:1.9.1-1ubuntu0.1) ...
Setting up git (1:1.9.1-1ubuntu0.1) ...
Setting up cgroup-lite (1.9) ...
cgroup-lite start/running
Setting up lxc-docker-1.7.1 (1.7.1) ...

Configuration file '/etc/init/docker.conf'
 ==> Deleted (by you or by a script) since installation.
 ==> Package distributor has shipped an updated version.
   What would you like to do about it ?  Your options are:
    Y or I  : install the package maintainer's version
    N or O  : keep your currently-installed version
      D     : show the differences between the versions
      Z     : start a shell to examine the situation
 The default action is to keep your current version.
*** docker.conf (Y/I/N/O/D/Z) [default=N] ? dpkg: error processing package lxc-docker-1.7.1 (--configure):
 EOF on stdin at conffile prompt
dpkg: dependency problems prevent configuration of lxc-docker:
 lxc-docker depends on lxc-docker-1.7.1; however:
  Package lxc-docker-1.7.1 is not configured yet.

dpkg: error processing package lxc-docker (--configure):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 lxc-docker-1.7.1
 lxc-docker
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

dmp42 commented:
@jasisk can you rebase against the latest master? Thanks!

@nathanleclaire wdyt?

nathanleclaire commented:
I'm hesitant, just because changing the shell commands in this domain could potentially be quite problematic.

@tianon Does it make sense to you?

tianon commented:
Hmm, I guess it makes sense that if you're asking docker-machine to manage the box that it should overwrite the config, but I'm not sure whether we should prefer "old" or "new" in this case.  I can see an argument for "old" because then a user can pre-configure a box, but the argument for "new" is that the user is specifically asking docker-machine to configure the box, so the default configuration would make sense to apply.

dgageot commented:
Hi @jasisk, can you rebase on master please?

jasisk commented:
Ack. Sorry, wasn't getting notifications for this one.

After working with this one for a while, I've changed my mind and have started using `force-confnew`. I ended up having more issues with confold` (particularly regarding different init systems), but `confined` has been without issue.

I can rebase this one if you want but would recommend switching to `confnew`.

dgageot commented:
@jasisk I'm not user I understand enough to give you an answer. @nathanleclaire, @tianon can you help decide?

nathanleclaire commented:
I think I would rather keep the user apt configuration as-is and try to detect an invalid state and warn if there is one (therefore I'm leaning on the side of closing the PR).  

@jasisk what causes the situation for you in the first place?  Did you pre-bake a machine image with `docker.conf` present but without Docker installed?  It will help to understand how the issue came up in the first place.

dgageot commented:
@jasisk @nathanleclaire What's the status of this PR? @jasisk I guess it should at least be rebased if you plan to get it merged. I'm still not sure I really understand the issue though.

jasisk commented:
Sorry, didn't get notifications again. I have to check my mail filters.

> what causes the situation for you in the first place? Did you pre-bake a machine image with docker.conf present but without Docker installed?

This was caused by a (very poorly) pre-provisioned base image for an openstack-based internal cloud. There exists no docker binary, nor `docker.conf` in `/etc/init` however `dpkg --get-selections` shows no less than three old versions (lxc-docker 1.3, 1.4, 1.5) in `deinstall`. I'm provisioning a brand new machine as we speak to check if there are other configs lying around (perhaps not upstart). In any case, my local solutions have been to force accept the new config or _actually_ purge the old configs (`aptitude purge ~c`).

> I guess it should at least be rebased if you plan to get it merged.

Noted. To follow shortly.

tianon commented:
Actually, I think https://github.com/docker/docker/issues/20524 is related here, specifically:

> See https://raphaelhertzog.com/2010/09/21/debian-conffile-configuration-file-managed-by-dpkg/, especially the section about "Avoiding the conffile prompt":
> 
>      apt-get -o Dpkg::Options::="--force-confnew" install -y ...

tianon commented:
I think `confnew` is probably a safer option for `docker-machine` to use, isn't it?

tianon commented:
(the rationale for that being that if I'm asking `docker-machine` to configure a machine for me, I want it to be configured by `docker-machine`)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "fix-conffile-error" git@github.com:jasisk/machine.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

jasisk commented:
> I think confnew is probably a safer option for docker-machine to use, isn't it?

Yup. That's what I landed on as well ([reference](#issuecomment-169200655)).

Rebased and changed to `confnew`.

nathanleclaire commented:
LGTM

nathanleclaire commented:
Thanks for the persistence @jasisk @tianon @dgageot

jasisk commented:
Likewise. Sorry for the long periods of silence. Fixed my email filters, fwiw. :grinning:
